### DISCUSSION TITLE
Unit Test Framework for cFS - Is there a user's guide or any documentation on this?

### DISCUSSION THREAD

**Discussion by motleekrugh** [nasa/cFS#741] [Q&A]
Is there anyway to understand more about the Unit Test Framework that is mentioned in the NASA documentation? I am not sure if there is a user's guide for this or if we need to build the cFS with a DOX parameter or exactly what? However, any help on understanding the current UTF for the 6.7 version of cFS is greatly appreciated. 

I would like to incorporate it in applications that are using the cFS and would like to incorporate the UTF into our applications and the more we know the better we understand to do this.

**Comment by avan989** [comment]
I have not seen any documentation but the sample app is a good start to understanding how to use the unit test framework. Anything in particular that you don't understand?

**Reply by motleekrugh** [reply]
The UTF in my mind at so many levels is not understood. This includes the overall concept of the UTF. How it is set up to best unit test an application. What is the overall scope of implementing stubs and hooks. All the different UT assert commands and examples of how to use them. I basically am looking for a user's guide related to the UTF. I know the number one answer related to all of these questions is to read the code. However, I just find that so unacceptable at so many different levels. We have gone through multiple implementations of the UTF on our applications and it is just so cumbersome and ineffective to do our unit testing in this method. There must be a better way to know how to implement the UTF and to best understand all the commands, the stubbing and the hooking of the code. The only write up I could find that provided some insights into this was an Analysis document on the effectiveness of unit testing the cFS. I was really hoping that this was documented somewhere and the official documentation could be provided to our development team so that we stop this flailing of trying to implement it correctly (based on the concepts and the design of how the UTF was supposed to be used with the applications).

**Reply by avan989** [reply]
I think you are looking for a "correct" way to do unit tests and there isn't a correct way. There might be a "current" way to do unit testing in the context of the open source, which is what you see in the "sample app"...etc. 

If you feel a standardized document is needed, I recommend putting in an issue ticket and possibly contributing to it as a first step but I feel this would be tough to maintain if the unit tests framework ever changes.
